syntax = "proto3";

package org.ojai.proto;

option java_multiple_files = true;
option java_package = "org.ojai.proto";

message OjaiKeyValue {
  string key = 1;
  OjaiValue value = 2;
}

message OjaiDocument {
  repeated OjaiKeyValue key_value = 1;
}

message OjaiList {
  repeated OjaiListItem value = 1;
}

message OjaiListItem {
  int32 index = 1;
  OjaiValue value = 2;
}

message OjaiValue {
  oneof data {
    bytes           null_value       =  1;
    bool            bool_value       =  2;
    string          string_value     =  3;
    sint32          byte_value       =  4;
    sint32          short_value      =  5;
    sfixed32        int_value        =  6;
    sfixed64        long_value       =  7;
    float           float_value      =  8;
    double          double_value     =  9;
    string          decimal_value    = 10;
    sint32          date_value       = 11;
    sint32          time_value       = 12;
    sint64          timestamp_value  = 13;
    sint64          interval_value   = 14;
    bytes           binary_value     = 15;
    OjaiDocument    map_value        = 16;
    OjaiList        list_value       = 17;
  }
}

// Protobuf encoded OJAI Objects
message DocumentPayload {
  repeated OjaiDocument documents = 1;
}
